src/main/java/games/strategy/triplea/attachments/TriggerAttachment.java

```diff
@@ -2144,38 +2140,34 @@ public static void triggerProductionFrontierEditChange(final Set<TriggerAttachme
       trigs = Matches.getMatches(trigs, availableUses);
     }
     final CompositeChange change = new CompositeChange();
-    for (final TriggerAttachment t : trigs) {
-      if (testChance && !t.testChance(bridge)) {
+    for (final TriggerAttachment triggerAttachment : trigs) {
+      if (testChance && !triggerAttachment.testChance(bridge)) {
         continue;
       }
       if (useUses) {
-        t.use(bridge);
-      }
-      final Iterator<String> iter = t.getProductionRule().iterator();
-      while (iter.hasNext()) {
-        boolean add = true;
-        final String[] s = iter.next().split(":");
-        final ProductionFrontier front = data.getProductionFrontierList().getProductionFrontier(s[0]);
-        String rule = s[1];
-        if (rule.startsWith("-")) {
-          rule = rule.replaceFirst("-", "");
-          add = false;
-        }
-        final ProductionRule productionRule = data.getProductionRuleList().getProductionRule(rule);
-        if (add) {
+        triggerAttachment.use(bridge);
+      }
+      triggerAttachment.getProductionRule().stream().map(s -> s.split(":")).forEach(array -> {
```

Comment:
There probably should be a new unit test to confirm that there is no change in behavior between the two implementations.  It's a bit complex for me to do in my head. :smile:

Reply:
I agree it's a bit complex, but writing a Unit Test is going to be hard: I have absolutely no idea what this code is trying to achieve.
If it helps: There are 3 steps involved here:
1. Replace the iterator while loop with a mapped stream forEach call
2. Remove the add boolean which is the same as `!rule.startsWith("-")` and make the rule string final
3. Rename vars

Reply:
One improvement for readability might be to put the `.map()` and `.forEach()` calls on separate lines.  The diff (even when ignoring whitespace changes) is already pretty busy, so I don't think the additional indentation is going to make it much worse.  It's just that I got hung up for about 30 seconds thinking the `forEach()` call was embedded in the `map()` lambda.

Reply:
> I agree it's a bit complex, but writing a Unit Test is going to be hard: I have absolutely no idea what this code is trying to achieve.

When I'm confronted with a situation like this, I write characterization tests for the original code.  That is, I don't really know what the results _should_ be, I just accept that whatever the original code outputs is correct.  Then I at least have a "vise" in place to prevent my refactoring from changing the existing behavior that I've characterized.

Reply:
Viewing this as a split diff rather than a unified diff made it much easier to process.  It all makes sense now. :smile:

I actually think having the intermediate Boolean made it a bit more readable, otherwise you have to read past the `"-"` processing to understand what's going on in each branch, but I'm ok with the code as it is.

FWIW, I was thinking of something like the following:

```diff
       triggerAttachment.getProductionRule().stream().map(s -> s.split(":")).forEach(array -> {
         final ProductionFrontier front = data.getProductionFrontierList().getProductionFrontier(array[0]);
         final String rule = array[1];
-        if (!rule.startsWith("-")) {
-          final ProductionRule productionRule =
-              data.getProductionRuleList().getProductionRule(rule);
+        final String ruleName = rule.replaceFirst("^-", "");
+        final ProductionRule productionRule = data.getProductionRuleList().getProductionRule(ruleName);
+        final boolean ruleAdded = !rule.startsWith("-");
+        if (ruleAdded) {
           if (!front.getRules().contains(productionRule)) {
             change.add(ChangeFactory.addProductionRule(productionRule, front));
             bridge.getHistoryWriter().startEvent(MyFormatter.attachmentNameToText(triggerAttachment.getName()) + ": "
                 + productionRule.getName() + " added to " + front.getName());
           }
         } else {
-          final ProductionRule productionRule =
-              data.getProductionRuleList().getProductionRule(rule.replaceFirst("-", ""));
           if (front.getRules().contains(productionRule)) {
             change.add(ChangeFactory.removeProductionRule(productionRule, front));
             bridge.getHistoryWriter().startEvent(MyFormatter.attachmentNameToText(triggerAttachment.getName()) + ": "
```